llvm/lib/Transforms/Yk/LivenessAnalysis.cpp

```diff
@@ -79,17 +79,23 @@ LivenessAnalysis::LivenessAnalysis(Function *Func) {
       if (!I.getType()->isVoidTy())
         Defs[&I].insert(cast<Value>(&I));
 
-      // Record what this instruction uses.
+      // For normal instructions we just iterate over all operands and mark
+      // them as used. We can't do this for PHI nodes though, since depending
+      // on the control flow (i.e. which block we are coming from) only one of
+      // the operands is used at a time. For example:
       //
-      // In order to track the operands of PHI nodes we need to be a bit crafty
-      // as otherwise we end up with live values in blocks where they actually
-      // don't exist. To avoid this, we need to mark as live any PHI operand
-      // from one block at the end of all other blocks. This leads to those
-      // values being killed immediately after entering the block in the
-      // backwards pass and thus the value never being live there.
+      //   phi [%1, bb1], [%2, bb2], [%3, bb3]
+      //
+      // Here, when the previous block was bb1, then only %1 is used, %2 if we
+      // come from bb2, and so on.
+      //
+      // The next problem is that we can't mark the operand as used in the
+      // current block, since that use would perculate upwards when looking at
+      // the predecessor blocks, and thus consider the operand used in blocks
+      // where it isn't. Instead, we want to mark the operand as used
+      // only in the predecssor block: %1 is marked as used at the end of bb1,
+      // %2 at the end of bb2, and so on.
```

Comment:
Let's credit Chris Lattner for this simple fix and link to the mailing list post?

Reply:
I thought about it, but I don't think it adds anything, and I don't like reddit links in comments.

Reply:
Not even the mailing list link?

Reply:
The important info in that link is just a single sentence: "PHI uses occur in the predecessor blocks, not in the PHI node blocks".  The rest doesn't seem too relevant: https://lists.llvm.org/pipermail/llvm-dev/2003-November/000542.html

Reply:
OK, I see you feel strongly, so as you wish.